File: VB/ObjectDataSourceInCode/Sample1.vb
Imports DevExpress.XtraReports.UI
Imports System.Collections.Generic
Imports System.Linq

Namespace ObjectDataSourceInCode.Sample1
    Public Class CodeSample

        ' This sample demonstrates how to bind a report to the data that an object's property exposes
        Public Shared Function CreateReport() As XtraReport

            Dim objectDS = New DevExpress.DataAccess.ObjectBinding.ObjectDataSource() With { _
                .Name = "ObjectDataSource1", _
                .Constructor = New DevExpress.DataAccess.ObjectBinding.ObjectConstructorInfo(), _
                .DataMember = "Items", _
                .DataSource = GetType(DataModel) _
            }

            Dim report = New XtraReport() With { _
                .Bands = { _
                    New TopMarginBand() With {.Name = "TopMargin"}, _
                    New DetailBand() With { _
                        .Name = "Detail", _
                        .Controls = { _
                            New XRLabel() With { _
                                .Name = "label1", .ExpressionBindings = { New ExpressionBinding("Text", "[Name]") }, _
                                .Size = New System.Drawing.Size(100, 23) _
                            } _
                        } _
                        _
                    }, _
                    New BottomMarginBand() With {.Name = "BottomMargin"} _
                }, _
                .DataSource = objectDS _
            }

            Return report
        End Function
    End Class
    Public Class DataModel
        Public Property Items() As List(Of DataItem)
        Public Sub New()
            Items = New List(Of DataItem)(Enumerable.Range(3, 5).Select(Function(item) New DataItem() With {.Name = "Name " & item}))
        End Sub
    End Class
    Public Class DataItem
        Public Property Name() As String
    End Class
End Namespace

File: CS/ObjectDataSourceInCode/Sample1.cs
using DevExpress.XtraReports.UI;
using System.Collections.Generic;
using System.Linq;

namespace ObjectDataSourceInCode.Sample1 {
    public class CodeSample {

        // This sample demonstrates how to bind a report to the data that an object's property exposes
        public static XtraReport CreateReport() {

            var objectDS = new DevExpress.DataAccess.ObjectBinding.ObjectDataSource() {
                Name = "ObjectDataSource1",
                Constructor = new DevExpress.DataAccess.ObjectBinding.ObjectConstructorInfo(),
                DataMember = "Items",
                DataSource = typeof(DataModel),
            };

            var report = new XtraReport() {
                Bands = {
                    new TopMarginBand() { Name = "TopMargin" },
                    new DetailBand() {
                        Name = "Detail",
                        Controls = {
                            new XRLabel() { 
                                Name = "label1",
                                ExpressionBindings = { new ExpressionBinding("Text", "[Name]") },
                                Size = new System.Drawing.Size(100, 23),
                            }
                        }
                    },
                    new BottomMarginBand() { Name = "BottomMargin" },
                },
                DataSource = objectDS
            };

            return report;
        }
    }
    public class DataModel {
        public List<DataItem> Items { get; set; }
        public DataModel() {
            Items = new List<DataItem>(
                Enumerable.Range(3, 5).Select(item => new DataItem() { Name = "Name " + item })
            );
        }
    }
    public class DataItem {
        public string Name { get; set; }
    }
}
